sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,535 @@
+ * https://cs140e.sergio.bz/docs/BCM2837-ARM-Peripherals.pdf
+ */
+
+#include <sys/mimiker.h>
+#include <sys/rman.h>
+#include <sys/devclass.h>
+#include <sys/klog.h>
+#include <sys/types.h>
+#include <dev/bcm2835_gpio.h>
+#include <sys/device.h>
+#include <sys/bus.h>
+#include <sys/time.h>
+#include <sys/libkern.h>
+#include <sys/callout.h>
+#include <sys/condvar.h>
+#include <sys/bus.h>
+#include <dev/emmc.h>
+#include <sys/errno.h>
+#include <sys/bitops.h>
+#include <dev/bcm2835_emmcreg.h>
+
+typedef struct bcmemmc_state {
+  resource_t *gpio;      /* GPIO resource (needed until we have a decent
+                          * way of setting up GPIO) */
+  resource_t *emmc;      /* e.MMC controller registers */
+  resource_t *irq;       /* e.MMC controller interrupt */
+  condvar_t intr_recv;   /* Used to wake up a thread waiting for an interrupt */
+  spin_t lock;           /* Covers `pending`, `intr_recv` and `emmc`. */
+  uint64_t rca;          /* Relative Card Address */
+  uint64_t host_version; /* Host specification version */
+  volatile uint32_t pending; /* All interrupts received */
+} bcmemmc_state_t;
+
+#define b_in bus_read_4
+#define b_out bus_write_4
+#define b_clr(mem, reg, b) b_out((mem), (reg), b_in((mem), (reg)) & ~(b))
+#define b_set(mem, reg, b) b_out((mem), (reg), b_in((mem), (reg)) | (b))
+
+/**
+ * Delay function.
+ * This just loops `count` times in a way that the compiler
+ * won't optimize away.
+ * \param count number of cycles to delay
+ */
+static void delay(int64_t count) {
+  __asm__ volatile("1: subs %[count], %[count], #1; bne 1b"
+                   : [ count ] "+r"(count));
+}
+
+static inline uint32_t emmc_wait_flags_to_hwflags(emmc_wait_flags_t mask) {
+  return ((mask & EMMC_I_DATA_DONE) ? INT_DATA_DONE : 0) |
+         ((mask & EMMC_I_READ_READY) ? INT_READ_RDY : 0) |
+         ((mask & EMMC_I_WRITE_READY) ? INT_WRITE_RDY : 0);
+}
+
+/* Returns new pending interrupts.
+ * All pending interrupts are stored in `bmcemmc_state_t::pending`. */
+static uint32_t bcmemmc_read_intr(bcmemmc_state_t *state) {
+  resource_t *emmc = state->emmc;
+  uint32_t newpend = b_in(emmc, BCMEMMC_INTERRUPT);
+  /* Pending interrupts need to be cleared manually. */
+  if (newpend) {
+    state->pending |= newpend;
+    b_out(emmc, BCMEMMC_INTERRUPT, newpend);
+  }
+  return newpend;
+}
+
+#define BCMEMMC_TIMEOUT 10000 /* Timeout when awaiting an interrupt */
+
+/**
+ * \brief Wait for all specified interrupts.
+ * \param dev eMMC device
+ * \param mask expected interrupts
+ * \param clear additional interrupt bits to be cleared
+ * \return 0 on success, EIO on internal error, ETIMEDOUT on device timeout.
+ */
+static int bcmemmc_intr_wait(device_t *dev, uint32_t mask) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+
+  assert(mask != 0);
+
+  SCOPED_SPIN_LOCK(&state->lock);
+
+  for (;;) {
+    (void)bcmemmc_read_intr(state);
+
+    if ((state->pending & mask) == mask) {
```

Comment:
I wonder if the interrupt processing makes sense now. Are we ever going to wait for error interrupts? Perhaps we should check them first. Is it a good idea that we clear `pending` on error – perhaps we should leave it for further analysis, before they're zeroed out?

Reply:
I redone the interrupt handling code and now errors are checked first.

This also revealed an earlier overlooked issue where SD card driver fails to send ACMD41 during a short polling phase when initializing a card. The exact error is a timeout on command line and it's actually an expected behavior. With that in mind I've introduced an `ERROR` property to e.MMC interface which lets the user check last error flags triggered when awaiting for interrupts (this includes sending commands as the controller is awaiting interrupt for every command sent).

See 3991851427074f3c9af6950ff1f39b68b6eafda7 update of the SD Card driver (#1135) for context.